### PR TITLE
CIFAR100 multilabel + mean/std

### PR DESCRIPTION
**Pull request by lucasb-eyer**
This should not change any behaviour by default, but allows to optionally include the coarse labels in the CIFAR100 dataset.

Adding the mean/std as attribute to the dataset is a little nicer than hard-coding them in the examples or needing everyone using this to re-compute them.

### CODE REVIEW

**Comment by szagoruyko** [comment]
we should allow grabbing raw data from datasets to use for other purposes, this PR breaks this by renaming `train_data` etc to `data`, and breaks a lot my code.

**Comment by lucasb-eyer** [comment]
Ah, bummer. In that case, I should close this PR since I'm also changing the data-structure of `self.{train_,test_,}labels`, which would also break your code.

Alternatively, I could do the following in this PR at the end of the constructor, but it's kinda ugly:

```python
# For backward compatibility:
if self.train:
    self.train_data = self.data
    self.train_labels = self.labels['labels']
else:
    # etc.
```

Let me know what you'd prefer (closing, or this "fix").

**Comment by szagoruyko** [comment]
@lucasb-eyer tbh I am not happy with the difference in naming `train_data` and `test_data`, maybe it's better to change it to `data` and `labels` after all.

**Comment by lucasb-eyer** [comment]
Nice we agree! I amended the last commit to fix the linter failures, anything else to get this merged?

In case there is a depreciation mechanism in PyTorch, I could still add the aliases and mark them deprecated.

**Comment by szagoruyko** [comment]
@lucasb-eyer we're at beta stage, I think we can just rename it, no need for deprecation yet.

File: torchvision/datasets/cifar.py [hunk 1]
```diff
@@ -31,6 +31,11 @@ class CIFAR10(data.Dataset):
         ['test_batch', '40351d587109b95175f43aff81a1287e'],
     ]
 
+    label_keys = ['labels']
```
Comment (alykhantejani) [inline torchvision/datasets/cifar.py]: do we not need `fine_labels` in here too?

**Comment by pqn** [comment]
@lucasb-eyer Friendly ping, I would love to see this merged.

File: torchvision/datasets/cifar.py (hunk 1 above)
Comment (lucasb-eyer) [reply torchvision/datasets/cifar.py]: No, there are no `fine_labels` in CIFAR10, only `labels`, see the old code.

**Comment by lucasb-eyer** [comment]
Oh, I completely forgot about this, I haven't used CIFAR100 for quite a while now.

Anyways, I went ahead and rebased it to the latest master, fixed conflicts, and hopefully got things right. Unfortunately I can't test it right now because my home cuda/cudnn/pytorch setup is currently broken, so I'd be happy if you can give it a try.

File: torchvision/datasets/cifar.py [hunk 2]
```diff
@@ -176,3 +160,14 @@ class CIFAR100(CIFAR10):
     test_list = [
         ['test', 'f0ef6b0ae62326f3e7ffdfab6717acfc'],
     ]
+
```
Comment (pqn) [inline torchvision/datasets/cifar.py]: Insert `label_keys = ['fine_labels']`

File: torchvision/datasets/cifar.py [hunk 3]
```diff
@@ -176,3 +160,14 @@ class CIFAR100(CIFAR10):
...
+    mean = [129.30416561, 124.0699627, 112.43405006]
+    std = [68.1702429, 65.39180804, 70.41837019]
+
+    def __init__(self, root, train=True,
+                 transform=None, target_transform=None,
+                 download=False, include_coarse=False):
+        if include_coarse:
+            self.label_keys = ['fine_labels', 'coarse_labels']
```
Comment (pqn) [inline torchvision/datasets/cifar.py]: Change to `self.label_keys.append('coarse_labels')`

**Comment by lucasb-eyer** [comment]
@pqn: No, these changes would be subtly wrong, as you can see from this example code:

```
>>> class Foo:
...   a = [1,2]
...   def __init__(self, b):
...     self.a.append(b)
...   def foo(self):
...     return self.a
... 
>>> f1 = Foo(10)
>>> f2 = Foo(20)
>>> f1.foo()
[1, 2, 10, 20]
>>> f2.foo()
[1, 2, 10, 20]
```

**Comment by pqn** [comment]
@lucasb-eyer Ah, I've deleted the incorrect suggestion, but the first one should fix things.

```
>>> class Foo:
...     a = ["foo"]
...     def __init__(self, b):
...         if b:
...             self.a = ["foo", "bar"]
...     def p(self):
...         print(self.a)
...
>>> f1 = Foo(False); f1.p()
['foo']
>>> f2 = Foo(True); f2.p()
['foo', 'bar']
>>> f3 = Foo(False); f3.p()
['foo']
>>> f1.p()
['foo']
```

**Comment by lucasb-eyer** [comment]
Only now do I see why you suggested that in the first place! (I missed the fine-only case for CIFAR100.) I have fixed it in the last commit, and now that I have a working PyTorch environment again, I gave it a quick one-off test and it worked for me.

**Comment by pqn** [comment]
@alykhantejani How does this look now?

File: torchvision/datasets/cifar.py [hunk 4]
```diff
@@ -62,44 +66,23 @@ def __init__(self, root, train=True,
...
-            fo = open(file, 'rb')
-            if sys.version_info[0] == 2:
-                entry = pickle.load(fo)
-            else:
-                entry = pickle.load(fo, encoding='latin1')
-            self.test_data = entry['data']
-            if 'labels' in entry:
-                self.test_labels = entry['labels']
-            else:
-                self.test_labels = entry['fine_labels']
```
Comment (alykhantejani) [inline torchvision/datasets/cifar.py]: @lucasb-eyer This case is no longer handled for cifar10 then? as now it only checks for `entry['labels']` If I understand the changes correctly.

**Comment by alykhantejani** [comment]
@pqn generally looks good to merge, I just have one question about `fine_labels` for the `CIFAR10` testing set which I just commented on the PR. If that's ok then its good to merge

File: torchvision/datasets/cifar.py (hunk 4 above)
Comment (lucasb-eyer) [reply torchvision/datasets/cifar.py]: I'm not sure I understand the question. CIFAR10 only has `labels`, there is no such thing as `fine_labels` and `coarse_labels` in CIFAR10.

The removed code you are commenting on only used `labels` in CIFAR10 (which still is used after the changes) and only used `fine_labels` in CIFAR100 (which now is either `fine_labels`, `coarse_labels`, or both).

Not sure if this answers your question.
